### ISSUE TITLE
Make it easier to set the container's default lifestyle

### ISSUE THREAD

**Issue by dotnetjunkie** [simpleinjector/SimpleInjector#54]
Since v2.6, Simple Injector allows users to override the way a lifestyle is determined for a registration. Simple Injector will register a component with the Transient lifestyle in case the user doesn't explicitly supply a lifestyle to one of the `Register` overloads, but the `ILifestyleSelectionBehavior` abstraction allows the user to determine a lifestyle on a per-type basis, as explained [here](https://simpleinjector.readthedocs.org/en/latest/extensibility.html#overriding-selection-behavior) in the documentation.

The `ILifestyleSelectionBehavior` abstraction is especially useful in case the user needs to select a lifestyle on a per-type basis (for instance by marking types with attributes). Although this abstraction is easy to use and implement, it still gives some overhead in case don't need a per-type override, but simply want all registrations to be made with a particular lifestyle, such as singleton.

To be able to do this, you would need the following code:

``` c#
public class SingletonLifestyleSelectionBehavior : ILifestyleSelectionBehavior {
    public Lifestyle SelectLifestyle(Type serviceType, Type implementationType) {
        return Lifestyle.Singleton;
    }
}

var container = new Container();
container.Options.LifestyleSelectionBehavior = new SingletonLifestyleSelectionBehavior();
```

Although this is okay, it would be much more convenient for the user to simply do this:

``` c#
var container = new Container();
container.Options.DefaultLifestyle = Lifestyle.Singleton;
```

Alternatively, a custom `ILifestyleSelectionBehavior` implementation could also be included in the framework, for instance:

``` c#
public class DefaultLifestyleSelectionBehavior : ILifestyleSelectionBehavior {
    public readonly Lifestyle DefaultLifestyle;
    public DefaultLifestyleSelectionBehavior(Lifestyle defaultLifestyle) {
        Requires.IsNotNull(defaultLifestyle, "defaultLifestyle");
        this.DefaultLifestyle = defaultLifestyle;
    }
    public Lifestyle SelectLifestyle(Type serviceType, Type implementationType) {
        return defaultLifestyle;
    }
}
```

This would allow the user to do this:

``` c#
var container = new Container();
container.Options.LifestyleSelectionBehavior = 
    new DefaultLifestyleSelectionBehavior(Lifestyle.Singleton);
```

Advantage of this is that we don't have two separate mechanisms to influence the container's default lifestyle (that might even conflict, because the `Options.DefaultLifestyle` will not be used when the `Options.LifestyleSelectionBehavior` is overridden), but on the other hand is such a `DefaultLifestyleSelectionBehavior` class less discoverable than having an `Options.DefaultLifestyle` property.

**Comment by TheBigRic** [comment]
IMO there should be one option to control this. The LifestyleSelectionBehaviour is more flexible and more in line with options as ConstructorSelectionBehaviour, so I favor this above a single option for singleton or transient.

**Comment by dotnetjunkie** [comment]
So should we add a `DefaultLifestyleSelectionBehavior` to the API then?

**Comment by qujck** [comment]
What will be the default setting for `container.Options.LifestyleSelectionBehavior`?

**Comment by dotnetjunkie** [comment]
When a user calls one of the `Register` overloads that doesn't specify a lifestyle, the container will call back into the `Options.LifestyleSelectionBehavior`. In its default configuration, it simply returns `Lifestyle.Transient` when called. This behavior will not change in v3.

**Comment by qujck** [comment]
I agree with Ric, it sits neatly alongside items such as `ConstructorSelectionBehavior`.

**Comment by qujck** [comment]
I wonder if this is this something best left to the documentation? If real world use cases arise then it is always something to add later.

E.g. how likely is it there will be anything other than a `SingletonLifestyleSelectionBehavior`? For example is `ExecutionContextLifestyleSelectionBehavior` something that would ever be considered a good idea? And if it was then it would probably be part of the relevant NuGet package.

Do you have any other use cases>

**Comment by dotnetjunkie** [comment]
Good point. I really can't imagine the default lifestyle to be replaced to anything else than Singleton. My idea was to add a `DefaultLifestyleSelectionBehavior` or something similar, that takes a `Lifestyle` in the constructor. This allows any lifestyle to be used, but there might not be a real common use case for that. So perhaps a `SingletonLifestyleSelectionBehavior` would be better. Especially with the coming area of the Singleton object graph.

Adding something like a `SingletonLifestyleSelectionBehavior` however is only useful if it is a real common use case and it communicates best practice. This is however, as you might know, something I'm really starting to believe. Object graphs should be singleton.

**Comment by qujck** [comment]
Until `SingletonLifestyleSelectionBehavior` is demonstrated as a workable best practice should it be left out?

**Comment by TheBigRic** [comment]
To throw in a completely different approach. Could container.options.LifestyleSelctionBehaviour be used for doing really complex things with lifestyles, by providing context to SelectLifestyle() method? Therby removing the need for Lifestyle.Hybrid? Just a qujck thought ;-)

**Comment by dotnetjunkie** [comment]
@TheBigRic, the `LifestyleSelectionBehavior` already allows determining the lifestyle of a component by analyzing both the service type by which it is registered and the actual implementation type. This allows a custom `LifestyleSelectionBehavior` implementation for instance to use the metadata (such as attributes) that is defined on such implementation type to determine the correct lifestyle.

Going one step further, as you suggest, and determining the lifestyle based on the context, seems like a dangerous idea to me. Or at least, I think it is dangerous to allow users to select the lifestyle based on contextual information of the consuming component just as we now do with the `RegisterConditional`. A component would always have a preferred lifestyle, and changing the lifestyle based on its consumer, means that the same component can exist in the same object graph with multiple lifestyles. This seems counterintuitive and error prone and I don't think there is a good use case for this. In case someone wants to register the same component with two different lifestyles, it probably means he is violating SRP and that component needs to be split up in two separate components. For that reason the [Ambiguous Lifestyles](https://simpleinjector.readthedocs.org/en/latest/ambiguouslifestyles.html) diagnostic warning exist and enabling such feature would immediately cause this warning to go off.

The predicate that is supplied to a `Hybrid` lifestyle is called (at least once) every time an object graph is resolved that contains components that are registered with this graph. This allows to decide at runtime which lifestyle should be used. This differs from the context used in `RegisterConditional`, which only allows compile-time decisions. Although you might say that the `Hybrid` lifestyle allows switching the lifestyle, but it will typically be used to mix scoped lifestyles, and this actually only switches the cache to look in to resolve an instance. Different contexts need different caches. A web request caches instances per `HttpContext`, while a background process typically uses a thread-local field. Of course the `Hybrid` lifestyle can be used to mix `Singleton` and `Transient` lifestyles with other lifestyles, but this is not a use case that is very plausible (and it is arguable whether this should actually be allowed) and it still doesn't allow to change the lifestyle based on parent information.

Long story short, I don't think that supplying contextual information to the `LifestyleSelectionBehavior` is a good idea, and since `LifestyleSelectionBehavior` is a compile-time feature while the `CreateHybrid` allows runtime decision making, they both have different purposes.

**Comment by dotnetjunkie** [comment]
Considering the discussion above, the most sensible thing would be to add a `SingletonLifestyleSelectionBehavior` to Simple Injector. However, adding such class give quite some overhead for us, because it needs to be tested, documented and XML documented, while it would take a user three lines of code to do this himself.

For this reason I am closing this issue; we will not add this to v3.
